velox/connectors/hive/storage_adapters/s3fs/S3FileSystem.cpp

```diff
@@ -0,0 +1,283 @@
+
+// Implement the S3ReadFile
+class S3ReadFile final : public ReadFile {
+ public:
+  S3ReadFile(std::string path, std::shared_ptr<Aws::S3::S3Client> client)
+      : client_(client) {
+    getBucketAndKeyFromS3Path(path, bucket_, key_);
+  }
+
+  // gets the length of the file
+  // checks if there are any issues accessing the file
+  void initialize() {
+    // make it a no-op if invoked twice.
+    if (length_ != -1) {
+      return;
+    }
+
+    Aws::S3::Model::HeadObjectRequest req;
+    req.SetBucket(getAwsString(bucket_));
+    req.SetKey(getAwsString(key_));
+
+    auto outcome = client_->HeadObject(req);
+    VELOX_CHECK_AWS_OUTCOME(
+        outcome, "failed to initialize S3 file", bucket_, key_);
+    length_ = outcome.GetResult().GetContentLength();
+    VELOX_CHECK(length_ >= 0);
+  }
+
+  std::string_view pread(uint64_t offset, uint64_t length, Arena* arena)
+      const override {
+    char* pos = arena->reserve(length);
+    preadInternal(offset, length, pos);
+    return {pos, length};
+  }
+
+  std::string_view pread(uint64_t offset, uint64_t length, void* buf)
+      const override {
+    preadInternal(offset, length, static_cast<char*>(buf));
+    return {static_cast<char*>(buf), length};
+  }
+
+  std::string pread(uint64_t offset, uint64_t length) const override {
+    // TODO: use allocator that doesn't initialize memory?
+    std::string result(length, 0);
+    char* pos = result.data();
+    preadInternal(offset, length, pos);
+    return result;
+  }
+
+  uint64_t preadv(
+      uint64_t offset,
+      const std::vector<folly::Range<char*>>& buffers) override {
+    VELOX_NYI();
+  }
+
+  uint64_t size() const override {
+    return length_;
+  }
+
+  uint64_t memoryUsage() const override {
+    // TODO: Check if any buffers are being used by the S3 library
+    return sizeof(Aws::S3::S3Client) + kS3MaxKeySize + 2 * sizeof(std::string) +
+        sizeof(int64_t);
+  }
+
+  bool shouldCoalesce() const final {
+    return false;
+  }
+
+ private:
+  // The assumption here is that "pos" has space for at least "length" bytes
+  void preadInternal(uint64_t offset, uint64_t length, char* pos) const {
+    // Read the desired range of bytes
+    Aws::S3::Model::GetObjectRequest req;
+    Aws::S3::Model::GetObjectResult result;
+
+    req.SetBucket(getAwsString(bucket_));
+    req.SetKey(getAwsString(key_));
+    std::stringstream ss;
+    ss << "bytes=" << offset << "-" << offset + length - 1;
+    req.SetRange(getAwsString(ss.str()));
+    // TODO: Avoid copy below by using  req.SetResponseStreamFactory();
+    // Reference: ARROW-8692
+    auto outcome = client_->GetObject(req);
+    VELOX_CHECK_AWS_OUTCOME(
+        outcome, "failure in S3ReadFile::preadInternal", bucket_, key_);
+
+    result = std::move(outcome).GetResultWithOwnership();
+    auto& stream = result.GetBody();
+    stream.read(reinterpret_cast<char*>(pos), length);
+  }
+  std::shared_ptr<Aws::S3::S3Client> client_;
+  std::string bucket_;
```

Comment:
const for bucket_ and key_

Reply:
these two variables are initialized via a non-const function call in the initiator. So can not be const.